pkg/ipam/node_manager_test.go

```diff
@@ -695,3 +734,171 @@ func BenchmarkAllocDelay50Worker10(b *testing.B) {
+		preExistingMaxAboveWm int
+		expPreAllocate        int
+		expMinAllocate        int
+		expMaxAllocate        int
+		expMaxAboveWatermark  int
+	}{
+		{
+			name:                 "instance-limits-under-operator-values",
+			nodeName:             "node-limits-under-operator",
+			expPreAllocate:       customPreAllocate,       // 8 - Operator value used (under limit of 10)
+			expMinAllocate:       customMinAllocate,       // 10 - Operator value used (at limit of 10)
+			expMaxAllocate:       customMaxAllocate,       // 20 - Operator value used (under limit of 30)
+			expMaxAboveWatermark: customMaxAboveWatermark, // 5 - use operator value
+		},
+		{
+			name:                 "instance-limits-over-operator-values",
+			nodeName:             "node-limits-over-operator",
+			expPreAllocate:       4,                       // Limited by minInstanceLimit=4 (operator value is 8)
+			expMinAllocate:       4,                       // Limited by minInstanceLimit=4 (operator value is 10)
+			expMaxAllocate:       15,                      // Limited by maxInstanceLimit=15 (operator value is 20)
+			expMaxAboveWatermark: customMaxAboveWatermark, // 5 - use operator value
+		},
+		{
+			name:                  "pre-existing-values-preserved",
+			nodeName:              "node-default",
+			preExistingValues:     true,
+			preExistingPreAlloc:   12,
+			preExistingMinAlloc:   14,
+			preExistingMaxAlloc:   25,
+			preExistingMaxAboveWm: 7,
+			expPreAllocate:        12, // Pre-existing value preserved
+			expMinAllocate:        14, // Pre-existing value preserved
+			expMaxAllocate:        25, // Pre-existing value preserved
+			expMaxAboveWatermark:  7,  // Pre-existing value preserved
+		},
+		{
+			name:                 "different-min-max-instance-limits",
+			nodeName:             "node-different-min-max",
+			expPreAllocate:       4,                       // Limited by minInstanceLimit=4 (operator value is 8)
+			expMinAllocate:       4,                       // Limited by minInstanceLimit=4 (operator value is 10)
+			expMaxAllocate:       20,                      // Operator value of 20 (under limit of 50)
+			expMaxAboveWatermark: customMaxAboveWatermark, // 5 - use operator value
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			node := &v2.CiliumNode{
+				ObjectMeta: metav1.ObjectMeta{Name: tc.nodeName},
+				Spec: v2.NodeSpec{
+					IPAM: ipamTypes.IPAMSpec{},
+				},
+			}
+
+			if tc.preExistingValues {
+				node.Spec.IPAM.PreAllocate = tc.preExistingPreAlloc
+				node.Spec.IPAM.MinAllocate = tc.preExistingMinAlloc
+				node.Spec.IPAM.MaxAllocate = tc.preExistingMaxAlloc
+				node.Spec.IPAM.MaxAboveWatermark = tc.preExistingMaxAboveWm
+			}
+
+			mngr.Upsert(node)
+
+			upsertedNode := mngr.Get(node.GetName())
+			require.NotNil(t, upsertedNode)
+
+			// Trigger the k8s sync to ensure syncToAPIServer is called
+			upsertedNode.k8sSync.Trigger()
+
+			time.Sleep(100 * time.Millisecond)
```

Comment:
We should avoid these kinds of sleeps for a specified period, in particular in this code as it relies on a lot of triggers which have caused tricky timing/race issues in the past.  Instead I would maybe do an `require.Eventually` (or similar) on `k8sapi.Get(node.GetName())` and proceed only when error is nil. Maybe with a second or two as the timeout. That should be more robust.